### PR TITLE
UW 575: Add shave driver

### PR DESCRIPTION
**Pull request by WeirAE**
**Synopsis**

Adds the shave driver, along with its tests, tested JSONschema, and documentation.

Rendered docs are here: https://uwtools--480.org.readthedocs.build/en/480/

**Type**

- [ ] Bug fix (corrects a known issue)
- [ ] Code maintenance (refactoring, etc. without behavior change)
- [ ] Documentation
- [x] Enhancement (adds a new functionality)
- [ ] Tooling (CI, code-quality, packaging, revision-control, etc.)

**Impact**

- [ ] This is a breaking change (changes existing functionality)
- [x] This is a non-breaking change (existing functionality continues to work as expected)

**Checklist**

- [x] I have added myself and any co-authors to the PR's _Assignees_ list.
- [x] I have reviewed the documentation and have made any updates necessitated by this change.

### CODE REVIEW

File: src/uwtools/tests/api/test_drivers.py [hunk 1]
```diff
@@ -16,13 +16,25 @@
...
+    chgres_cube,
+    esg_grid,
+    fv3,
+    jedi,
+    make_hgrid,
+    mpas,
+    mpas_init,
+    sfc_climo_gen,
+    ungrib,
+    shave,
```
Comment (elcarpenterNOAA) [inline src/uwtools/tests/api/test_drivers.py:35]:
```suggestion
    shave,
    ungrib,
```

File: docs/index.rst [hunk 2]
```diff
@@ -133,6 +133,12 @@ sfc_climo_gen
 | **CLI**: ``uw sfc_climo_gen -h``
 | **API**: ``import uwtools.api.drivers.sfc_climo_gen``
 
+shave
+"""""
+
+| **CLI**: ``uw shave -h``
+| **API**: ``import uwtools.api.drivers.shave``
```
Comment (maddenp-noaa) [inline docs/index.rst:140]:
```suggestion
| **API**: ``import uwtools.api.shave``
```
I'm fixing this on my UPP branch for the existing drivers, but we might as well fix this one before it merges.

File: docs/sections/user_guide/cli/drivers/shave.rst [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+       runscript
+         The runscript
+       validate
+         Validate the UW driver config
+
+All tasks take the same arguments. For example:
+
+.. code-block:: text
+
+   $ uw shave run -h
```
Comment (maddenp-noaa) [inline docs/sections/user_guide/cli/drivers/shave.rst:36]:
```suggestion
   $ uw shave run --help
```

File: docs/sections/user_guide/cli/drivers/shave.rst [hunk 4]
```diff
@@ -0,0 +1,93 @@
+``shave``
+=========
+
+The ``uw`` mode for configuring and running the UFS Utils preprocessing component ``shave``. Documentation for this UFS Utils component is :ufs-utils:`here <shave>`.
+
+.. code-block:: text
+
+   $ uw shave -h
```
Comment (maddenp-noaa) [inline docs/sections/user_guide/cli/drivers/shave.rst:8]:
```suggestion
   $ uw shave --help
```
For consistency with most of the other docs. There are a couple of exceptions that I'm fixing on my branch.

File: docs/sections/user_guide/cli/drivers/shave.rst [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+Examples
+^^^^^^^^
+
+The examples use a configuration file named ``config.yaml`` with content similar to:
+
+.. highlight:: yaml
+.. literalinclude:: ../../../../shared/shave.yaml
+
+
```
Comment (maddenp-noaa) [inline docs/sections/user_guide/cli/drivers/shave.rst:70]:
```suggestion

```

File: docs/sections/user_guide/yaml/components/shave.rst [hunk 6]
```diff
@@ -0,0 +1,46 @@
+.. _shave_yaml:
+
+shave
+========
```
Comment (maddenp-noaa) [inline docs/sections/user_guide/yaml/components/shave.rst:4]:
```suggestion
=====
```

File: src/uwtools/resources/jsonschema/shave.jsonschema [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+              "type": "integer"
+            },
+            "nx": {
+              "type": "integer"
+            },
+            "ny": {
+              "type": "integer"
+            }
+          }
+        },
```
Comment (maddenp-noaa) [inline src/uwtools/resources/jsonschema/shave.jsonschema:22]: Are all of these optional?

File: docs/sections/user_guide/cli/drivers/shave.rst [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+   Optional arguments:
+     -h, --help
+         Show help and exit
+     --version
+         Show version info and exit
+
+   Positional arguments:
+     TASK
+       input_file
+         Ensure the specified input grid file exists
```
Comment (maddenp-noaa) [inline docs/sections/user_guide/cli/drivers/shave.rst:22]: I don't see such a task.

File: docs/shared/shave.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+shave:
+  execution:
+    batchargs:
+      cores: 1
+      walltime: "00:01:00"
+    executable: /path/to/shave_nc.F90
```
Comment (maddenp-noaa) [inline docs/shared/shave.yaml:6]: "Is the executable actually named with a .F90 extension?", he asked, sadly expecting the answer "Yes".

File: src/uwtools/resources/jsonschema/shave.jsonschema (hunk 7 above)
Comment (WeirAE) [reply src/uwtools/resources/jsonschema/shave.jsonschema:22]: Ah, I went through a few versions and looks like I didn't re-require them in the end, thanks! I'll add that to the test too

File: docs/shared/shave.yaml (hunk 9 above)
Comment (WeirAE) [reply docs/shared/shave.yaml:6]: Oh interesting - actually, it turns out that they did not do that! There's a CMakeLists in the directory that compiles it to just "shave", let me fix this

File: src/uwtools/tests/test_schemas.py [hunk 10]
```diff
@@ -774,6 +779,37 @@ def test_schema_sfc_climo_gen_run_dir(sfc_climo_gen_prop):
...
+            "nh4": 1,
+        },
+        "execution": {"executable": "shave"},
+        "run_dir": "/tmp",
+    }
+    errors = schema_validator("shave", "properties", "shave")
+    # Basic correctness:
+    assert not errors(config)
+    # Some top-level keys are required:
+    for key in ("execution", "run_dir", "config"):
```
Comment (NaureenBharwaniNOAA) [inline src/uwtools/tests/test_schemas.py:800]:
```suggestion
    for key in ("config", "execution", "run_dir"):
```

File: src/uwtools/tests/test_schemas.py [hunk 11]
```diff
@@ -774,6 +779,37 @@ def test_schema_sfc_climo_gen_run_dir(sfc_climo_gen_prop):
...
+        assert f"'{key}' is a required property" in errors(with_del(config, key))
+    # Additional top-level keys are not allowed:
+    assert "Additional properties are not allowed" in errors({**config, "foo": "bar"})
+
```
Comment (NaureenBharwaniNOAA) [inline src/uwtools/tests/test_schemas.py:807]: You could add a test here to confirm that the values under `config` need to be of a specific type as well.

Similar to what exists for `esg-grid`.

i.e. 
```
assert "not of type 'number'" in errors({"config": {"nx": "/some/str"}})
```

**Review by maddenp-noaa — APPROVED** [review]

**Review by NaureenBharwaniNOAA — APPROVED** [review]
Nice work!

File: src/uwtools/tests/test_schemas.py [hunk 12]
```diff
@@ -774,6 +779,56 @@ def test_schema_sfc_climo_gen_run_dir(sfc_climo_gen_prop):
...
+    for key in ("config", "execution", "run_dir"):
+        assert f"'{key}' is a required property" in errors(with_del(config, key))
+    # Additional top-level keys are not allowed:
+    assert "Additional properties are not allowed" in errors({**config, "foo": "bar"})
+
+
+def test_schema_shave_config_properties():
+    # Get errors function from schema_validator
+    errors = schema_validator("shave", "properties", "shave", "properties", "config")
+    for key in ("input_grid_file", "nx", "ny", "nh4"):
```
Comment (NaureenBharwaniNOAA) [inline src/uwtools/tests/test_schemas.py:809]: Nice set of tests!

**Review by elcarpenterNOAA — APPROVED** [review]
Nice work!

**Review by maddenp-noaa — APPROVED** [review]

**Pull request merged**